### PR TITLE
Add more objects to templates_and_annotations

### PR DESCRIPTION
**Pull request by ToxicKidz**
## Description
This PR adds the role, user, member, and emoji objects, along with `UserClientManager.users`, `Guild.emojis`, `Guild.roles`, and `Guild.members`. This also adds the `MemberTemplate` that will be implemented later.

### CODE REVIEW

File: snakecord/clients/user/manager.py [hunk 1]
```diff
@@ -119,37 +149,110 @@ class guild_integrations_update(_base_event):
     @eventdef
     @dataclass
     class guild_member_add(_base_event):
-        pass
+        guild: Optional[Guild] = None
+        member: Optional[GuildMember] = None
+        
+        def __post_init__(self):
+            guild_id = self.payload.pop('guild_id')
```
Comment (asleep-cult) [inline snakecord/clients/user/manager.py]: `.get('guild_id')`

File: snakecord/objects/base.py [hunk 2]
```diff
@@ -5,8 +5,11 @@
 from ..utils.json import JsonObject
 from ..utils.snowflake import Snowflake
 
+__all__ = ('BaseObject',)
 
 class BaseObject(JsonObject, template=BaseTemplate):
+    __slots__ = ('id',)
```
Comment (asleep-cult) [inline snakecord/objects/base.py]: This is implicitly added

File: snakecord/objects/user.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+class User(BaseObject, template=UserTemplate):
+    __slots__ = ('_state',)
+
+    def __init__(self, *, state: UserState):
+        self._state = state
+    
+    @property
+    def mention(self):
+        return f'<@!{self.id}>'
```
Comment (asleep-cult) [inline snakecord/objects/user.py]: The ! is only for guild members with nicknames

File: snakecord/states/base.py [hunk 4]
```diff
@@ -91,6 +91,12 @@ def find(self, func):
 
     def append(self, data: dict):
         raise NotImplementedError
+    
+    def extend(self, data: List[dict]) -> list:
```
Comment (asleep-cult) [inline snakecord/states/base.py]:
```py
def extend(self, data: Iterable[dict]) -> list:
    return [self.append(d) for d in data]
```

File: snakecord/clients/user/manager.py [hunk 5]
```diff
@@ -119,37 +149,110 @@ class guild_integrations_update(_base_event):
...
     class guild_member_remove(_base_event):
-        pass
+        guild: Optional[Guild] = None
+        user: Optional[User] = None
+        member: Optional[GuildMember] = None
+
+        def __post_init__(self):
+            guild_id = self.payload.get('guild_id')
+            user = self.payload.get('user')
+            self.user = self.manager.users.append(user)
```
Comment (asleep-cult) [inline snakecord/clients/user/manager.py:180]: Handle user in `GuildMemberState.append` instead

File: snakecord/states/guild.py [hunk 6]
```diff
@@ -7,16 +7,15 @@ class GuildState(BaseState):
...
         cls.__guild_class__ = klass
 
     def append(self, data: dict, *args, **kwargs) -> Guild:
         guild = self.get(data['id'])
         if guild is not None:
             guild._update(data)
         else:
-            guild = self.__guild_class__.unmarshal(data, state=self, *args,
-                                                   **kwargs)
+            guild = self.__guild_class__.unmarshal(data, state=self, *args, **kwargs)
```
Comment (asleep-cult) [inline snakecord/states/guild.py]: flake8

File: snakecord/templates/message.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    webhook_id=JsonField('webhook_id'),
+    type=JsonField('type'),
+    _activity=JsonField('activity'),
+    application=JsonField('application'),
+    _message_reference=JsonField('message_reference'),
+    flags=JsonField('flags'),
+    _stickers=JsonArray('stickers'),
+    _referenced_message=JsonField('referenced_message'),
+    _interaction=JsonField('interaction'),
+    __extends__=(BaseTemplate,)
```
Comment (asleep-cult) [inline snakecord/templates/message.py:31]: This shouldn't extend BaseTemplate

File: snakecord/states/user.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+from .base import BaseState, SnowflakeMapping
+from ..objects.user import User
+
+class UserState(BaseState):
+    __container__ = SnowflakeMapping
+    __user_class__ = User
+
+    @classmethod
+    def set_user_class(cls, class_: type) -> None:
```
Comment (asleep-cult) [inline snakecord/states/user.py]: call this `klass` instead of `class_`

**Review by asleep-cult — CHANGES_REQUESTED** [review]
Run through flake8

File: snakecord/templates/message.py (hunk 7 above)
Comment (ToxicKidz) [reply snakecord/templates/message.py:31]: Why not?

Comment (asleep-cult) [reply snakecord/templates/message.py:31]: BaseTemplate is for things with an id

Comment (ToxicKidz) [reply snakecord/templates/message.py:31]: Messages have IDs

File: snakecord/clients/user/manager.py (hunk 5 above)
Comment (asleep-cult) [reply snakecord/clients/user/manager.py:180]: I mean in `guild_member_update`

File: snakecord/templates/message.py (hunk 7 above)
Comment (asleep-cult) [reply snakecord/templates/message.py:31]: thought this was embed for some reason

**Pull request merged**